Source/Core/Core/HW/GCMemcard/GCMemcard.cpp

```diff
@@ -27,196 +27,256 @@ static void ByteSwap(u8* valueA, u8* valueB)
+{
+  GCMemcard card;
+  card.m_filename = std::move(filename);
+
+  // TODO: Format() not only formats the card but also writes it to disk at m_filename.
+  // Those tasks should probably be separated.
+  if (!card.Format(shift_jis, size_mbits))
+    return std::nullopt;
+
+  return std::move(card);
+}
+
+std::pair<GCMemcardErrorCode, std::optional<GCMemcard>> GCMemcard::Open(std::string filename)
+{
+  GCMemcardErrorCode error_code;
+  File::IOFile file(filename, "rb");
+  if (!file.IsOpen())
   {
-    PanicAlertT("Failed to read 1st block allocation table block correctly\n(0x6000-0x7FFF)");
-    return;
+    error_code.Set(GCMemcardValidityIssues::FAILED_TO_OPEN);
+    return std::make_pair(error_code, std::nullopt);
   }
 
-  if (!mcdFile.ReadBytes(&m_bat_blocks[1], BLOCK_SIZE))
+  // check if the filesize is a valid memory card size
+  u16 card_size_mbits;
+  const u64 filesize = file.GetSize();
+  switch (filesize)
   {
-    PanicAlertT("Failed to read 2nd block allocation table block correctly\n(0x8000-0x9FFF)");
-    return;
+  case MegabitsToBytes(MBIT_SIZE_MEMORY_CARD_59):
+    card_size_mbits = MBIT_SIZE_MEMORY_CARD_59;
+    break;
+  case MegabitsToBytes(MBIT_SIZE_MEMORY_CARD_123):
+    card_size_mbits = MBIT_SIZE_MEMORY_CARD_123;
+    break;
+  case MegabitsToBytes(MBIT_SIZE_MEMORY_CARD_251):
+    card_size_mbits = MBIT_SIZE_MEMORY_CARD_251;
+    break;
+  case MegabitsToBytes(MBIT_SIZE_MEMORY_CARD_507):
+    card_size_mbits = MBIT_SIZE_MEMORY_CARD_507;
+    break;
+  case MegabitsToBytes(MBIT_SIZE_MEMORY_CARD_1019):
+    card_size_mbits = MBIT_SIZE_MEMORY_CARD_1019;
+    break;
+  case MegabitsToBytes(MBIT_SIZE_MEMORY_CARD_2043):
+    card_size_mbits = MBIT_SIZE_MEMORY_CARD_2043;
+    break;
+  default:
+    error_code.Set(GCMemcardValidityIssues::INVALID_CARD_SIZE);
+    return std::make_pair(error_code, std::nullopt);
   }
 
-  u32 csums = TestChecksums();
-
-  if (csums & 0x1)
+  // read the entire card into memory
+  GCMemcard card;
+  file.Seek(0, SEEK_SET);
+  if (!file.ReadBytes(&card.m_header_block, BLOCK_SIZE) ||
+      !file.ReadBytes(&card.m_directory_blocks[0], BLOCK_SIZE) ||
+      !file.ReadBytes(&card.m_directory_blocks[1], BLOCK_SIZE) ||
+      !file.ReadBytes(&card.m_bat_blocks[0], BLOCK_SIZE) ||
+      !file.ReadBytes(&card.m_bat_blocks[1], BLOCK_SIZE))
   {
-    // header checksum error!
-    // invalid files do not always get here
-    PanicAlertT("Header checksum failed");
-    return;
+    error_code.Set(GCMemcardValidityIssues::IO_ERROR);
+    return std::make_pair(error_code, std::nullopt);
   }
 
-  if (csums & 0x2)  // 1st directory block checksum error!
+  const u16 card_size_blocks = card_size_mbits * MBIT_TO_BLOCKS;
+  const u16 user_data_blocks = card_size_blocks - MC_FST_BLOCKS;
+  card.m_data_blocks.reserve(user_data_blocks);
+  for (u16 i = 0; i < user_data_blocks; ++i)
   {
-    if (csums & 0x4)
+    card.m_data_blocks.emplace_back();
```

Comment:
Since C++17 this returns the inserted element; can we just use it for clarity?
```c++
auto& new_data_block = card.m_data_blocks.emplace_back();
if (!file.ReadArray(new_data_block.m_block.data(), BLOCK_SIZE))
```
Or are there any compilers that still don't fully support C++17?

Reply:
I didn't realize that was a thing! Yeah, that's definitely nicer.